Clear sensitive memory without getting optimized out

real-or-random opened the pull request:
This is a rebase of #448, including a few changes (mostly implementing what I described in #185). I haven't tested this on Windows and I actually haven't reviewed this in detail so far, but people can start to have a look.

Fixes #185 and closes #448.

real-or-random commented on src/util.h:
```diff
@@ -125,6 +131,32 @@ static SECP256K1_INLINE void *manual_alloc(void** prealloc_ptr, size_t alloc_siz
+#if defined(_MSC_VER)
+    /* SecureZeroMemory is guaranteed not to be optimized out by MSVC. */
+    SecureZeroMemory(ptr, n);
+#elif defined(__GNUC__)
```
I couldn't figure out when GCC started to support assembly but it's some version <= 2.95... So if you have such an old compiler, I guess you're on your own.

gmaxwell replied:
GCC 2.95 is likely the earliest we'd would ever encounter someone wanting to support. (Haiku OS uses it still, I believe :) )

real-or-random commented on src/util.h:
```diff
@@ -11,10 +11,16 @@
 
+#if defined(_MSC_VER)
+// For SecureZeroMemory
+#include <Windows.h>
```
I think that's okay but I'm not entirely sure if we want that. It's used in Bitcoin Core too.

gmaxwell replied:
I would be that this should be version gated, I'll see if I can figure out where thats supported to.

gmaxwell commented:
I think using a sizeof on an argument is likely to introduce almost impossible to detect bugs when e.g. someone hands it the first element of an array (esp one passed from another function).  It doesn't feel particularly typesafe to me.

Is there a reason why you preferred this approach to having every subsystem (like field, scalar) define a clear function for its relevant type?

real-or-random commented:
No, I don't prefer that to be honest. It's just the approach taken in #448, and I was somewhat worried about it, too.

real-or-random commented:
I removed the macro from the PR. Also I removed `_fe_set_zero` from the PR (one can use `_fe_set_int`).

I'm happy to take suggestions for additional code locations where memory should be cleared.

real-or-random commented on src/util.h:
```diff
@@ -125,6 +131,31 @@ static SECP256K1_INLINE void *manual_alloc(void** prealloc_ptr, size_t alloc_siz
+     * just not remove it entirely. See "Dead Store Elimination (Still) Considered Harmful" by
+     * Yang et al. (USENIX Security 2017) for more background.
+     */
+    memset(ptr, 0, len);
```
Now that "clearing" is cleanly separated from "setting to zero", it's a good idea to replace 0 with a different value here for testing that the code does not rely on the fact that the "_clear" functions set to 0. We could actually always use a different value here, or use a different value in (some runs of) tests. What do people think about this?

elichai replied:
Did you try checking the binary that it's actually not optimized out?

real-or-random replied:
Not yet. But yes, we should do this with a few compilers.

jonasnick replied:
> it's a good idea to replace 0 with a different value here for testing

I don't think it's a good idea to use a different value in testing. But I don't see a downside to just replace `0` with `42`. This also has the (slight) advantage that it'll make it easier to detect now improper usage of `*_clear` when rebasing -zkp on top of this. -zkp uses `*_clear` for initializing in many places.

real-or-random commented on src/field.h:
```diff
@@ -49,10 +49,11 @@ static int secp256k1_fe_normalizes_to_zero(secp256k1_fe *r);
-/** Set a field element equal to a small integer. Resulting field element is normalized. */
+/** Set a field element equal to a small integer. Resulting field element is normalized; it has
+ *  magnitude 0 if a == 0, and magnitude 1 otherwise. */
 static void secp256k1_fe_set_int(secp256k1_fe *r, int a);
```
(I wondered why this accepts `int`. An unsigned type seems more appropriate. For type safety, we would even want to exclude too large values and use something as small as `unsigned char`. But I guess this is slower and just not worth the hassle)?

elichai commented:
What about clearing the memory of the temporary FE in the GE addition? https://github.com/bitcoin-core/secp256k1/blob/master/src/group_impl.h#L419 (and I think most of this file)

real-or-random commented:
Thanks, collecting more:
https://github.com/bitcoin-core/secp256k1/blob/master/src/scalar_impl.h#L71
A lot in https://github.com/bitcoin-core/secp256k1/blob/master/src/scalar_8x32_impl.h  (also also 4x64)

And I think there more in the field module... I guess I'll just go through the entire codebase. :shrug:

gmaxwell commented:
@elichai  *_var are not used on secret inputs and probably should not waste time zeroizing anything.

jonasnick commented:
Results are positive and very similar to the above with clang version 8.0.0.

elichai commented:
@jonasnick did you try to benchmark if this affects performance in a non negligible way?

jonasnick commented:
On my system the memory barrier and volatile function pointer method have no measurable performance difference. However, using one of the methods vs. using nothing in `bench_sign` appears to be `1.2%` slower (`41.3us` vs `40.8us`, averaged over 6 runs).

jonasnick commented:
Looks like memsetting to `0x42` instead of `0` costs about 1% :/
```
ecdsa_sign: min 41.3us / avg 41.5us / max 42.2us
vs.
ecdsa_sign: min 40.9us / avg 41.1us / max 41.8us
```
(I changed run_benchmark count argument from 10 to 500)

sipa commented:
@jonasnick That sounds like a deal breaker.

real-or-random commented:
Note that's just signing time, not verification time.

But I agree, let's stick to 0, that's just simpler.

sipa commented:
Concept ACK.

There are a number of data types which end up being cleared using the generic `secp256k1_mem_clear()` function after this PR (including at least `secp256k1_sha256` and `secp256k1_rfc6979_hmac_sha256`). I think it would be cleaner to provide `*_clear()` functions for those as well, and use those where relevant.

real-or-random commented:
I rebased this and implemented @sipas suggestion for the hash API. Now the caller is responsible to call `_clean()` (instead of letting `_finalize`) do it. I think this is actually the right thing to do.



nit: Also renamed `secp256k1_mem_clear` to `memclear` now that we have also `memczero` in `util.h`

real-or-random commented:
> Going through the codebase to look for locations where a `_clear` is missing can happen in another PR.

Indeed, so this is ready for review.

For looking into more places in the future, in particular within some arithmetic function, I was wondering if "overdoing" this will introduce more stores to memory, which is not only slightly worse for security but will probably also affect performace. For example, the `explicit_bzero` function in Glibc has this problem:

> Also, explicit_bzero only operates on RAM. If a sensitive data object never needs to have its address taken other than to call explicit_bzero, it might be stored entirely in CPU registers until the call to explicit_bzero. Then it will be copied into RAM, the copy will be erased, and the original will remain intact. Data in RAM is more likely to be exposed by a bug than data in registers, so this creates a brief window where the data is at greater risk of exposure than it would have been if the program didn’t try to erase it at all. 

https://www.gnu.org/software/libc/manual/html_node/Erasing-Sensitive-Data.html

AFAIU and I've seen from looking at generated ASM this should not happen with our memory barrier approach but I'm not 100% sure and any insight will be helpful.

Note to myself: Cleaning registers is out of the scope currently but today the cited USENIX paper  reminded me of a simple but expensive way to do this: Just run the operation again. (For example, run another SHA256 transform on dummy midstate and dummy input.) This will reuse and hence overwrite the registers (modulo inlining etc).

real-or-random commented:
Apparently stores of 64-byte zero-over-zero can be faster than stores of other values, see https://twitter.com/BRIAN_____/status/1260913021116993536. This would be a reason not so overwrite with zeros.

But I think we shouldn't have secrets that are all zeroes.This would be relevant for indistinguishability-based primitives with arbitrary inputs, e.g.,  encryption, but not for signing and key exchange.

real-or-random commented:
> Conceptually I think it makes sense to have separate secure-erase wiping and zeroing of values. It needs a big redo by now though; are you still interesting in working on this?

Yeah, I certainly want to come back to this, but I don't want to commit to a specific timeline. Marking as draft for now.

siv2r commented:
Should this PR be closed since #1579 has replaced it?
